.changeset/brown-pets-clean.md

```diff
@@ -0,0 +1,7 @@
+---
+"astro": minor
+---
+
+The astro middleware now runs when a matching page or endpoint is not found. Previously, a `pages/404.astro` or `pages/[...catch-all].astro` route had to match to allow middleware. This is now not necessary.
+
+Note that some adapters may need to update to let Astro's SSR code handle the not-found case.
```

Comment:
nit: "to update.." what? It feels like we should hint at something or maybe rephrase the sentence. 

It would be great if the changelog actually explains what an adapter maintainer should update in order to "fix" this.

Reply:
I feel it adequately explains the change necessary.

Adapters may look at the list of routes given by astro to create their platform-specific routing file. And the platform may then run SSR code only for routes known to be on-demand rendered. The website would then fallback to a platform-designed 404 page. SST, vercel and cloudflare (in one routing strategy) work like this.